src/platform_impl/windows/dark_mode.rs

```diff
@@ -0,0 +1,222 @@
+                    dwMinorVersion: 0,
+                    dwBuildNumber: 0,
+                    dwPlatformId: 0,
+                    szCSDVersion: [0; 128],
+                };
+
+                (rtl_get_version)(&mut vi as _);
+
+                if vi.dwMajorVersion == 10 && vi.dwMinorVersion == 0 {
+                    Some(vi.dwBuildNumber)
+                } else {
+                    None
+                }
+            }
+        } else {
+            None
+        }
+    };
+
+    static ref DARK_MODE_SUPPORTED: bool = {
+        // We won't try to do anything for windows versions < 17763
+        // (Windows 10 October 2018 update)
+        match *WIN10_BUILD_VERSION {
+            Some(v) => v >= 17763,
+            None => false
+        }
+    };
+
+    static ref DARK_THEME_NAME: Vec<u16> = widestring("DarkMode_Explorer");
+    static ref LIGHT_THEME_NAME: Vec<u16> = widestring("");
+}
+
+/// Attempt to set dark mode on a window, if necessary.
+/// Returns true if dark mode was set, false if not.
+pub fn try_dark_mode(hwnd: HWND) -> bool {
+    if *DARK_MODE_SUPPORTED {
+        let is_dark_mode = should_use_dark_mode();
+
+        let theme_name = if is_dark_mode {
+            DARK_THEME_NAME.as_ptr()
+        } else {
+            LIGHT_THEME_NAME.as_ptr()
+        };
+
+        unsafe {
+            assert_eq!(
+                0,
+                uxtheme::SetWindowTheme(hwnd, theme_name as _, std::ptr::null())
+            );
+
+            set_dark_mode_for_window(hwnd, is_dark_mode)
+        }
+
+        is_dark_mode
+    } else {
+        false
+    }
+}
+
+fn set_dark_mode_for_window(hwnd: HWND, is_dark_mode: bool) {
+    // Uses Windows undocumented API SetWindowCompositionAttribute,
+    // as seen in win32-darkmode example linked at top of file.
+
+    type SetWindowCompositionAttribute =
+        unsafe extern "system" fn(HWND, *mut WINDOWCOMPOSITIONATTRIBDATA) -> BOOL;
+
+    #[allow(non_snake_case)]
+    type WINDOWCOMPOSITIONATTRIB = u32;
+    const WCA_USEDARKMODECOLORS: WINDOWCOMPOSITIONATTRIB = 26;
+
+    #[allow(non_snake_case)]
+    #[repr(C)]
+    struct WINDOWCOMPOSITIONATTRIBDATA
+    {
+        Attrib: WINDOWCOMPOSITIONATTRIB,
+        pvData: PVOID,
+        cbData: SIZE_T
+    }
+
+    lazy_static! {
+        static ref SET_WINDOW_COMPOSITION_ATTRIBUTE: Option<SetWindowCompositionAttribute> =
+            get_function!("user32.dll", SetWindowCompositionAttribute);
+    }
+
+    if let Some(set_window_composition_attribute) = *SET_WINDOW_COMPOSITION_ATTRIBUTE {
+        unsafe {
+            // SetWindowCompositionAttribute needs a bigbool (i32), not bool.
+            let mut is_dark_mode_bigbool = is_dark_mode as BOOL;
+
+            let mut data = WINDOWCOMPOSITIONATTRIBDATA {
+                Attrib: WCA_USEDARKMODECOLORS,
+                pvData: &mut is_dark_mode_bigbool as *mut _ as _,
+                cbData: std::mem::size_of_val(&is_dark_mode_bigbool) as _
```

Comment:
```suggestion
                cbData: std::mem::size_of_val(&is_dark_mode_bigbool).into()
```
The underlying type is `usize`.

Reply:
Unfortunately windows wants `u32` for the cbData value, so we need to do a narrowing cast here. Size of a bool is always going to fit safely in a `u32`, so this narrowing `as` isn't dangerous.